### PR TITLE
Feature/another map implementation

### PR DESCRIPTION
**Pull request by dwickstrom**
Hey there! I started on the same kind of idea, but luckily I didn't get too far until I found your work. Figured I should contribute here instead so here the first thing I had in mind: 

The idea was to add another implementation of map, that would allow mapping over types other than iterables, ie other functors too. Also added a few more function implementations available in ramda.

Let me know what you think 🎷

### CODE REVIEW

File: pyramda/prop.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from pyramda.function.curry import curry
+
+
+@curry
+def prop(x, obj):
+    if isinstance(obj, dict):
```
Comment (jackfirth) [inline pyramda/prop.py]: I don't think this should be either calling `getitem` or `getattr`, it should only use `getitem`. Unlike javascript, in python it's not common to treat objects and their attributes as dictionaries implicitly.

File: pyramda/range.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from pyramda.function.curry import curry
+
+arange = range
```
Comment (jackfirth) [inline pyramda/range.py]: You can `import builtins` and refer to range as `builtins.range` to avoid this renaming.

File: pyramda/prop.py (hunk 1 above)
Comment (dwickstrom) [inline pyramda/prop.py]: I'm a little hesitant on this. Before I remove the dictionary part maybe you could elaborate a little, along the lines of why that would be problematic or not so useful, or even aesthetically undesirable etc?

Comment (jackfirth) [inline pyramda/prop.py]: Certainly!

In Javascript, objects are frequently used both to represent traditional OOP-style objects with methods and encapsulated data and to represent traditional dictionaries with plain data in the form of key-value pairs. There is zero syntactic distinction between these two uses, dot-notation (`obj.key`) and bracket-notation (`obj['key']`) are equivalent.

This is absolutely not the case in Python, where dot-notation ("attribute access") is used for "objects" and bracket-notation ("item access") is used for "dictionaries". This is more of a loose rule, as it's possible to wrap both object-style objects and dictionary-style objects in proxies that flip the meaning of item access and attribute access. But this doesn't change the fact that, fundamentally, `obj.key` and `obj['key']` mean very, very different things in Python and the vast majority of Python objects will respond with completely different semantics.

As a result, I think that Pyramda should make a conscious effort not to mix up these two concepts. If a user wants to access both attributes and items with the same function, they can write a proxy class that overrides either `__getitem__` or `__getattr__` depending on how they want to access values.

Comment (dwickstrom) [inline pyramda/prop.py]: I like when I get to think about the fundamentals of a language for a bit. Thanks for taking the time to. Though about it for a moment and your reasoning is certainly more respectful towards Python itself, which is obviously a good idea.

So, I did notice there is already a `getattr` implementation in there and also a `getitem` in the `dictionary` folder, so I will remove the prop thing entirely :)

**Comment by dwickstrom** [comment]
Hi! Great, thanks for the feedback. I don't really have a lot of strong opinions on the dispatching, just sort of tasting different ideas at the moment. 

I did think of having an even more capable dispatcher for detecting more types of things that could be thought of as being functors, like sets, tuples and what not. And on the other side just checking if the value implements some specified (Fantasy land type spec) functor interface. Which to be fair, in a way is what I tried to do, perhaps just not in the most elegant way possible... I guess settling for either map/fmap (rather than aliasing one of them) could make things feel a bit more streamlined... The [fantasyland spec](https://github.com/fantasyland/fantasy-land#functor) uses `map` so...

I do have a hard time coming up with a better approach of detecting a functor that isn't really specified anywhere. Maybe contemplating on the matter for a bit helps resolving it 🌅

**Comment by dwickstrom** [comment]
Oh, I forgot to comment on the your idea of adding dispatching at a later stage (whole-hog approach). Don't have much to add here, but I did notice that the [PyMonad library uses the curry function to add some custom operator magic](https://bitbucket.org/jason_delaat/pymonad/src/cbecd6796cd1488237d2a0f057cefd2a50df753a/pymonad/Reader.py?at=master&fileviewer=file-view-default#Reader.py-91). Maybe the curry decorator is a good insertion point for some kind of type checking mechanism.

File: pyramda/prop.py (hunk 1 above)
Comment (jackfirth) [inline pyramda/prop.py]: Huh, I'd completely forgotten that `getitem` was already implemented in the first place. Oh well :)

**Comment by dwickstrom** [comment]
Did you think of doing a kind of type checking that would give an answer like this 
`get_type(val)` and then get an explicit type? I started thinking about this and made a bunch of methods like this:

```
has_method = curry(lambda method, obj: callable(getattr(obj, method, None)))
has_setoid_type = has_method('equals')
has_semigroup_type = has_method('concat')
has_monoid_type = compose(has_method('empty'), has_semigroup_type)
has_functor_type = has_method('map')
has_apply_type = compose(has_method('ap'), has_method('map'))
has_applicative_type = compose(has_method('of'), has_method('ap'), has_method('map'))
has_monad_type = compose(has_method('chain'), has_applicative_type)
has_foldable_type has_method('reduce')
has_traversable_type = compose(has_method('traverse'), has_foldable_type, has_functor_type)
has_chain_type = compose(has_method('chain'), has_apply_type)
has_chainrec_type = compose(has_method('chainrec'), has_chain_type)
has_monad_type = compose(has_chain_type, has_applicative_type)
has_extend_type = has_method('extend')
has_comonad_type = compose(has_extend_type, has_functor_type)
has_bifunctor_type = compose(has_method('bimap'), has_functor_type)
has_profunctor_type = compose(has_method('promap'), has_functor_type)
```

To be able to dispatch all of them in 1 method you would have to call a whole bunch of them to be able to give a "complete" answer. That might be a little too expensive...?

Maybe, it is actually better to let the functions themselves ask any type questions they need to? Like this:

```
@curry
def map(f, value):
    return value.map(f) if has_functor_type(value) else [f(x) for x in value]

@curry
def reduce(f, init, val):
    return val.reduce(f, init) if has_foldable_type else functools.reduce(f, val, init)

etc...
```

**Comment by jackfirth** [comment]
I would hold off on adding dispatching to this PR, that's a pretty huge piece to implement and I'd like it separated from the other functions included in here.

I agree with you, individual functions should decide how to dispatch rather than some decorator. I think there's a decent way to do this in a generic way with less boilerplate. Maybe something like this?

```python
@curry
def map(f, mapped):
  dispatcher = FunctorDispatcher(mapped)
  return dispatcher.map(f) if dispatcher else [f(v) for v in mapped]
```

...where `BlahDispatcher(x)` either returns false or an object that knows how to pass along to methods. This presents a nice place to encapsulate "fallback" logic, as something like `MonadDispatcher` could automatically derive an `fmap` method if the underlying value only provides a `bind` method.

A related discussion: Python has a convention of using `__foo__` for "secret" methods that the language hooks into for dispatching things like operator overloading. Python also recommends that "userland" hooks like this use single underscores, so a Fantasy Land project would probably dispatch to methods looking like `_fantastyland_map_`.

**Comment by dwickstrom** [comment]
Letting a dispatcher provide missing methods sounds a little too adventurous to me, but I'm not sure I get whole picture :)

On the topic of responsibilities - I'm thinking there might be no good reason to let the dispatcher do half the job. Isn't it better to let it do the whole thing?

```python
def map(f, context):
    return dispatch(f, context)
```

It seems a little half baked in splitting the responsibility between the map function and the dispatcher 

```python
def map(f, context):
    return context.map(f) if dispatch(context) else fallback
```
Perhaps just let the map function do the job, to actually provide the "know how", isn't such a bad idea?
```python
def map(f, context):
    return context.map(f) if this_is_a_functor(context) else fallback
```

Perhaps the map function is actually a great place to encapsulate specific fallback logic?

As you mentioned, a dispatcher could be a good place to encapsulate logic. I'd argue that this is the case if it can be generalised to some degree of usefulness. 

Knowing how to handle a potential functor in the case of the map function should maybe not even be regarded as boilerplate. Perhaps one would just end up in creating unnecessary indirections by moving this type validation to a central location, rather than adding value?

Anyhow, to move forward, do you suggest I remove the map function from this PR then, and save it for later?

**Comment by jackfirth** [comment]
Closing due to age.

**Comment by JBKahn** [comment]
I would also love to see fantasy land in pyramda. Specifically, I would love to see the fantasy land map exception.

**Pull request closed without merging**